Content.Server/Singularity/EntitySystems/SingularityAttractorSystem.cs

```diff
@@ -0,0 +1,106 @@
+    {
+        base.Initialize();
+        SubscribeLocalEvent<SingularityAttractorComponent, ComponentStartup>(OnAttractorStartup);
+    }
+
+    /// <summary>
+    /// Updates the pulse cooldowns of all singularity attractors.
+    /// If they are off cooldown it makes them emit an attraction pulse and reset their cooldown.
+    /// </summary>
+    /// <param name="frameTime">The time elapsed since the last set of updates.</param>
+    public override void Update(float frameTime)
+    {
+        if(!_timing.IsFirstTimePredicted)
+            return;
+
+        foreach(var (attractor, xform) in EntityManager.EntityQuery<SingularityAttractorComponent, TransformComponent>())
+        {
+            var curTime = _timing.CurTime;
+            if (attractor.NextPulseTime <= curTime)
+                Update(attractor.Owner, attractor, xform);
+        }
+    }
+
+    /// <summary>
+    /// Makes an attractor attract all singularities and puts it on cooldown.
+    /// </summary>
+    /// <param name="uid">The uid of the attractor to make pulse.</param>
+    /// <param name="attractor">The state of the attractor to make pulse.</param>
+    /// <param name="xform">The transform of the attractor to make pulse.</param>
+    private void Update(EntityUid uid, SingularityAttractorComponent? attractor = null, TransformComponent? xform = null)
+    {
+        if(!Resolve(uid, ref attractor))
+            return;
+
+        attractor.LastPulseTime = _timing.CurTime;
+        attractor.NextPulseTime = attractor.LastPulseTime + attractor.TargetPulsePeriod;
+        if (!Resolve(uid, ref xform))
+            return;
+
+        var mapPos = xform.Coordinates.ToMap(EntityManager);
+
+        if (mapPos == MapCoordinates.Nullspace)
+            return;
+
+        foreach(var (singulo, walk, singuloXform) in EntityManager.EntityQuery<SingularityComponent, RandomWalkComponent, TransformComponent>())
+        {
+            var singuloMapPos = singuloXform.Coordinates.ToMap(EntityManager);
+
+            if (singuloMapPos.MapId != mapPos.MapId)
+                continue;
+
+            var biasBy = mapPos.Position - singuloMapPos.Position;
+            if (biasBy.Length() <= MinAttractRange)
+                return;
+            biasBy = Vector2.Normalize(biasBy) * (attractor.BaseRange / biasBy.Length());
+
+            walk.BiasVector += biasBy;
+        }
+    }
+
+    /// <summary>
+    /// Resets the pulse timings of the attractor when the component starts up.
+    /// </summary>
+    /// <param name="uid">The uid of the attractor to start up.</param>
+    /// <param name="comp">The state of the attractor to start up.</param>
+    /// <param name="args">The startup prompt arguments.</param>
+    private void OnAttractorStartup(EntityUid uid, SingularityAttractorComponent comp, ComponentStartup args)
+    {
+        comp.LastPulseTime = _timing.CurTime;
+        comp.NextPulseTime = comp.LastPulseTime + comp.TargetPulsePeriod;
+    }
```

Comment:
This is going to bulldoze the data every time map is loaded which is incorrect, it should be done on MapInit.

Reply:
done, moved to MapInitEvent